[clang][PAC] add support for options parameter to __ptrauth

ojhunt opened the pull request:
[clang][PAC] add support for options parameter to __ptrauth

This PR adds support for an 'options' parameter for the __ptrauth
qualifier.

The initial version only exposes the authehntication modes:
 * "strip"
 * "sign-and-strip"
 * "sign-and-auth"

We also support parsing the options but not yet the implementation
 * "isa-pointer"
 * "authenticates-null-values"

The initial support for authentication mode controls exist to support
ABI changes over time, and as a byproduct support basic initial tests
for option parsing.

ojhunt commented:
The tests in this PR are dependent on #136828, and the included tests have non-sequential numbers as the tests include other options that are not supported by this PR, to try and limit the complexity of this PR to just the parsing and existing functionality.

ojhunt commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,191 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+        return;
+      }
+      OptionsString = OptionsStringLiteral->getString();
+    } else if (S.EvaluateAsString(AuthenticationOptionsArg, OptionsString, S.Context, Sema::StringEvaluationContext::PtrauthOptions, /*ErrorOnInvalidMessage=*/false)) {
```
The implementation I was upstreaming had a very very restricted version of @cor3ntin's constant evaluated string, so I have adopted that.

ojhunt commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,191 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+    bool HasSeenOption = false;
+    unsigned CurrentIdx = 0;
+
+    auto OptionStringIdxLocation = [&](unsigned Idx) {
```
This code is moderately concise parser of the grammar:

option :- [a-z\-]+
options :- option (',' option)*

The reason for the complexity - vs. just a split(',').map(trim) - is just error messaging. We could simplify if people think it's warranted, but the error message tests are already reduced from what it could be.

cor3ntin replied:
We could, if there is an error in the nth item, recompute the location by finding the nth comma and point to that. I feel like more granularity than that adds more complexity than it is worth.

ojhunt commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,191 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+      return std::make_pair(OptionStartIdx, OptionEndIdx);
+    };
+
+    auto OptionHandler = [&](StringRef TokenStr, SourceRange TokenRange,
```
This is more generic than is strictly needed at this point, because future PRs add additional option types and flags

cor3ntin replied:
I would much prefer a 3 if/else :)

ojhunt commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8350,14 +8350,16 @@ static void HandleNeonVectorTypeAttr(QualType &CurType, const ParsedAttr &Attr,
-         "__ptrauth qualifier takes between 1 and 3 arguments");
+  assert((Attr.getNumArgs() > 0 && Attr.getNumArgs() <= 4) &&
+         "__ptrauth qualifier takes between 1 and 4 arguments");
+  StringRef AttrName = Attr.getAttrName()->getName();
```
`AttrName` is used because the diagnostics need the actual attribute name, and there's a future PR to add the `__ptrauth_restricted_intptr` qualifier which means it is not a constant. We can remove `AttrName` now, but it is immediately needed in the next PR.

ojhunt replied:
following the __ptrauth for ints PR __ptrauth_restricted_intptr is gone, I thought I had resolved all of these

ojhunt commented:
This PR is a draft as it will need to be updated to merged correctly, and verify test output (conflicts, unintentional mismatches coming from splitting out the PR) after #136828

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,212 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+                                  S.Context,
+                                  Sema::StringEvaluationContext::PtrauthOptions,
+                                  /*ErrorOnInvalidMessage=*/false)) {
+      for (char Ch : OptionsString) {
+        if (!Ch || !isascii(Ch)) {
+          DiagnoseInvalidOptionsParameter(
+              "contains invalid characters", Ch,
+              AuthenticationOptionsArg->getSourceRange());
+          return;
+        }
+      }
```
Do we really need that? we can just say "\0签名" is not a valid option

cor3ntin replied:
Can you add tests with a string-view-like type? I would not be surprised if we need more changes in the attribute parsing code to be able to have string_view as `StringArgument` argument

ojhunt replied:
@cor3ntin re: string_view: I'm trying not to add anything beyond the bare minimum here, I thought the existing behavior of EvaluateString is a superset of what we know existing code requires so I didn't want to trying to add more complexity to a pile of already complex PRs, that said it looks like I haven't pulled those specific tests across so I could be wrong.

Re: non-ascii: I'm not sure I understand the semantics of StringLiteral::toString and EvaluateString(...) if presented w.r.t utf8 or wide chars so I wanted consistent behavior between them, and the easiest way to ensure that was an early out on non-ascii. StringLiteral::getString() specifically has an assertion specifically about expecting char sized characters which was worrying to me. If that assertion is actually trying to say sizeof(string code point)==sizeof(char) then that's much simpler.

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,212 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+      return;
+    }
+    if (OptionsStringLiteral) {
+      if (OptionsStringLiteral->containsNonAsciiOrNull()) {
+        DiagnoseInvalidOptionsParameter(
+            "contains invalid characters", std::nullopt,
+            AuthenticationOptionsArg->getSourceRange());
+        return;
+      }
```
Same here as below, if we accept exactly 3 strings, we don;t need to check anything else

ojhunt replied:
There are more than just those three strings, but adding the additional ones implies the ability to test them, which means pulling in the implementation of those options.

I went for this specific subset because PointerAuthQualifier already has the required functionality, and so this PR could be very close to just the parsing.

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8374,20 +8376,212 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+      Attr.setInvalid();
+      ReportEvaluationOfExpressionIfNeeded();
+    };
+    if (AuthenticationOptionsArg->isValueDependent() ||
+        AuthenticationOptionsArg->isTypeDependent()) {
+      DiagnoseInvalidOptionsParameter(
+          "is dependent", std::nullopt,
+          AuthenticationOptionsArg->getSourceRange());
+      return;
```
I wonder if that can just be delayed to later? Maybe a separate patch though

ojhunt replied:
I added this somewhat ad hoc, the behavior I was upstreaming basically just did `if (StringLiteral) { ... } else if (const eval produces a char* + length)  { ... }`, which just happens to exclude dependent values, but EvaluateString has an assertion that the values are not dependent. That might be something we want to fix in future, but it seemed that the sensible thing at the moment was just adding a dependent value/type check as we have no support for dependent ptrauth qualifiers :-/

ojhunt replied:
@cor3ntin oh, when you say delayed until later do you mean template instantiation? That would be nice but currently the ptrauth qualifier does not support dependent parameters at all - it's something I've wanted to add support for (because it would be very useful) but is never high enough on the priority list, and seems to require an inordinate amount of seemingly mechanical work to achieve :-/

ojhunt commented:
@cor3ntin the option selection down stream is a StringSwitch on the authentication mode, and then a sequence of (not large, I think just two) `if (option == constant)` checks to set up specific options.

Previous PRs have suggested maps, etc for such conditions, and a map of option=>handler seemed consistent with that in this case, but I'm really on the fence about which. Absent the map approach (just typing directly into the comment w/o checking code) I think the alternative is that the option handler looks like:

```cpp
switchswitch(option)
.Case("strip", ...)
.Case("sign-and-strip", ...)
.Case("sign-and-auth", ...);
// 'else' here I think is handled by a flag
else if (option == thing1) { .... }
else if (option == thing2) { ... }
```

Any thoughts on which you consider preferable? I think either works as the gross portion of this code is already just the tokenisation.

ojhunt commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8373,10 +8375,140 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+    } else {
+      Expr::EvalResult Eval;
+      bool Result = AuthenticationOptionsArg->EvaluateAsRValue(Eval, Ctx);
+      if (Result && Eval.Val.isLValue()) {
```
This section can be replaced once https://github.com/llvm/llvm-project/pull/135864 or similar is landed as it supports const char* constant expressions

ojhunt commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -1735,8 +1735,8 @@ def warn_pragma_unroll_cuda_value_in_parens : Warning<
 
-def err_ptrauth_qualifier_bad_arg_count : Error<
-  "'__ptrauth' qualifier must take between 1 and 3 arguments">;
+def err_ptrauth_qualifier_bad_arg_count
```
This is clang-format induced (aside from the 3 vs 4 change obviously) I can more the `Error<` back but it might be worth trying to work out how to stop clang-format from making this choice

AaronBallman replied:
I'd recommend manually formatting diagnostic files rather than using clang-format, at least until the tool learns (or can be configured for) our conventions.

ojhunt replied:
yeah, I wonder if we can make clang-format exclude the diagnostic files

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3595,9 +3595,9 @@ def ObjCRequiresPropertyDefs : InheritableAttr {
-              IntArgument<"ExtraDiscriminator", 1>];
+  let Args = [IntArgument<"Key">, BoolArgument<"AddressDiscriminated", 1>,
+              IntArgument<"ExtraDiscriminator", 1>,
+              StringArgument<"Options", 1>];
```
Should there be an update to AttrDocs.td for the new argument?

ojhunt replied:
I'll see if there's any existing documentation as the attribute can't be specified by attribute syntax

ojhunt replied:
Updated the PointerAuthentication document

AaronBallman commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -65,6 +65,17 @@ enum class PointerAuthenticationMode : unsigned {
   SignAndAuth
 };
 
+static constexpr llvm::StringLiteral PointerAuthenticationOptionStrip = "strip";
```
At the language option level, I think it's more idiomatic to use an enumeration instead of string literals and doing string processing.

erichkeane replied:
I'd suggest doing a string-switch on it, then storing an enum in the AST.  We can then print it/whatever on the way out.  But a pair of conversion functions is typically all you should need.  Though, it doesn't seem that these are leaving the `handle` function anyway, so IDK.

Either way, LangOpts is a little bit of an odd place for this to live.

ojhunt replied:
It's not stored as a string (or even an attribute), they're stored as flags and enums in the pointer auth qualifier on the type.

ojhunt replied:
I was meaning "is there a more idiomatic way to have these string constants specified?" :D

ojhunt replied:
Oh, I hadn't included the updated to stringifying the qualifier so these were only used in one place.

AaronBallman replied:
> I was meaning "is there a more idiomatic way to have these string constants specified?" :D

Usually we'd do what @erichkeane was suggesting -- not have these as constants in LangOptions.h but instead have helper functions which convert string -> enum -> string. Then these strings can be hidden within a .cpp file rather than exposed in a header with broader visibility. WDYT?

erichkeane reviewed (commented):
I think this is on the right path, but others have good comments here so far.

ojhunt commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -327,6 +330,27 @@ a discriminator determined as follows:
+separated modifiers to the normal authentication behavior. Currently supported
+options are
+
+- Authentication mode: This is one of ``strip``, ``sign-and-strip``, and
```
This PR does not actually provide the implementation of this option, as it touches a number of places unrelated to parsing the option.

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -327,6 +330,27 @@ a discriminator determined as follows:
+- Authentication mode: This is one of ``strip``, ``sign-and-strip``, and
+  ``sign-and-auth``. The ability to modify this behavior is intended to support
+  staging ABI changes. The ``strip`` mode results in the PAC bits of a value
+  being stripped from any value and disabled any other authentication
```
```suggestion
  being stripped from any value and disabling any other authentication
```

AaronBallman commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -327,6 +330,27 @@ a discriminator determined as follows:
+  ensure a correct signature is set on assignment. Finally ``sign-and-auth`` is
+  the default mode, and provides full protection for the value.
+
+- ``authenticates-null-values``: By default the __ptrauth qualifier does not
```
```suggestion
- ``authenticates-null-values``: By default the ``__ptrauth`` qualifier does not
```

ojhunt committed:
[clang][PAC] add support for options parameter to __ptrauth

This PR adds support for an 'options' parameter for the __ptrauth
qualifier.
The initial version only exposes the authehntication modes:
 * "strip"
 * "sign-and-strip"
 * "sign-and-auth"

These modes exist to support ABI migration, and should not be used
beyond these specific use cases.

This also adds support for parsing the `authenticates-null-values`
and `isa-pointer` options, but does not introduce codegen changes
yet.

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -1048,6 +1048,19 @@ def err_ptrauth_extra_discriminator_invalid : Error<
+// __ptrauth qualifier options string
+def note_ptrauth_evaluating_options
+    : Note<"options parameter evaluated to '%0'">;
+def err_ptrauth_invalid_option : Error<"'__ptrauth' options parameter %0">;
```
Look for uses of `enum_select` in that file and do something similar here - lets avoid partial diagnostic strings in cpp files

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8479,10 +8480,131 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+
+    for (unsigned Idx = 0; Idx < Options.size(); ++Idx) {
+      StringRef OptionString = Options[Idx].trim();
+      auto HandleOption = [&](auto Value, auto *Option,
```
Can you use the actual type of the parameters, here? PointerAuthenticationMode afaict

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8479,10 +8480,131 @@ static void HandlePtrAuthQualifier(ASTContext &Ctx, QualType &T,
+              << OptionString.take_until(isWhitespace);
+
+        IsInvalid = true;
+      }
```
i think this whole loop would be a lot simpler if we looked first for end of the
current parameter (ie `find_if(OptionsString, [](char c) {return isWhitespace(c) || c == ','; }`) instead of splitting on `,` and then trying to decide if individual components are valid.

ojhunt commented:
*Converting to draft as I may be working on @apple-fcloutier's string constant PR which this will adopt.

cor3ntin commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -8631,17 +8632,162 @@ static void HandleNeonVectorTypeAttr(QualType &CurType, const ParsedAttr &Attr,
+  StringRef IsIsaPointerOption;
+  StringRef AuthenticatesNullValuesOption;
+  for (StringRef CurrentOption : Options) {
+    StringRef *Storage = nullptr;
```
There is no reason for this to be a pointer.
